Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,667 @@
+
+    public init(
+        body: @escaping sending (ByteBuffer, consuming LambdaResponseWriter, LambdaContext) async -> ()
+    )
+
+    public func handle(
+        _ request: ByteBuffer,
+        responseWriter: consuming LambdaResponseWriter,
+        context: LambdaContext
+    ) async
+}
+
+extension LambdaRuntime {
+    public init(
+        body: @escaping sending (ByteBuffer, consuming LambdaResponseWriter, LambdaContext) async -> ()
+    )
+}
+```
+
+## Alternatives considered
+
+### [UInt8] instead of ByteBuffer
+
+We considered using `[UInt8]` instead of `ByteBuffer` in the base `LambdaHandler` API. We decided to use `ByteBuffer`
+for two reasons.
+
+1. 99% of use-cases will use the JSON codable API and will not directly get in touch with ByteBuffer anyway. For those
+   users it does not matter if the base API uses `ByteBuffer` or `[UInt8]`.
+2. The incoming and outgoing data must be in the `ByteBuffer` format anyway, as Lambda uses SwiftNIO under the hood and
+   SwiftNIO uses `ByteBuffer` in its APIs. By using `ByteBuffer` we can save a copies to and from `[UInt8]`. This will
+   reduce the invocation time for all users.
+3. The base `LambdaHandler` API is most likely mainly being used by developers that want to integrate their web
+   framework with Lambda (examples: Vapor, Hummingbird, ...). Those developers will most likely prefer to get the data
+   in the `ByteBuffer` format anyway, as their lower level networking stack also depends on SwiftNIO.
+
+### Users create a LambdaResponse, that supports streaming instead of being passed a LambdaResponseWriter
+
+Instead of passing the `LambdaResponseWriter` in the invocation we considered a new type `LambdaResponse`, that users
+must return in the `StreamingLambdaHandler`.
+
+Its API would look like this:
+
+```swift
+/// A response returned from a ``LambdaHandler``.
+/// The response can be empty, a single ByteBuffer or a response stream.
+public struct LambdaResponse {
+    /// A writer to be used when creating a streamed response.
+    public struct Writer: ~Copyable {
+        /// Writes data to the response stream
+        public func write(_ byteBuffer: ByteBuffer) async throws
+        /// Closes off the response stream
+        public func finish() async throws
+        /// Writes the `byteBuffer` to the response stream and subsequently closes the stream
+        public func writeAndFinish(_ byteBuffer: ByteBuffer) async throws
+    }
+
+    /// Creates an empty lambda response
+    public init()
+
+    /// Creates a LambdaResponse with a fixed ByteBuffer.
+    public init(_ byteBuffer: ByteBuffer)
+
+    /// Creates a streamed lambda response. Use the ``Writer`` to send
+    /// response chunks on the stream.
+    public init(_ stream: @escaping sending (Writer) async throws -> ())
+}
+```
+
+The `StreamingLambdaHandler` would look like this:
+
+```swift
+/// The base LambdaHandler protocol
+public protocol StreamingLambdaHandler: ~Copyable {
```

Comment:
same question around allowing `~Copyable` types here.

Reply:
Let's discuss here: https://github.com/swift-server/swift-aws-lambda-runtime/pull/339#discussion_r1711515316 once we have a solution we will fix or resolve here.